### DISCUSSION TITLE
Is there a way to load and cache a WAV from SD? Or play from WAV stored in PROGMEM?

### DISCUSSION THREAD

**Discussion by UnexpectedMaker** [schreibfaul1/ESP32-audioI2S#629] [Q&A]
Hi folks - I want to load a few WAVS from SD and store them in memory, and play the from there, so I don't have to take the loading hit them every time to play them.

Think small SFX that get called at specific times to be played. The delay in loading them every time from SD is too much for what I am doing, and I can't seem to find a way to play a WAV from data stored in PROGMEM, or that gets loaded at runtime, if I load the data myself.

Am I missing something obvious here?

Thanks :)
Seon - Unexpected Maker

**Comment by codegrue** [comment]
I too am looking for a way to play from PROGMEM

**Comment by Marcus4242** [comment]
Did you ever figure out a solution to play a small sound file with minimum delay?

**Reply by codegrue** [reply]
I was never able to get this library to work. I ended up using this one instead:

https://github.com/pschatzmann/arduino-audio-tools

And this post has my exact code to achieve the PROGMEM playback:

https://github.com/pschatzmann/arduino-audio-tools/discussions/1337#discussioncomment-10606145

**Reply by Marcus4242** [reply]
Thanks for the update. I'll check it out 🙏

**Comment by schreibfaul1** [comment]
@Marcus4242 
This library is far too large and far too slow for your intended use due to the processes to be initiated.
@UnexpectedMaker
A file is always expected. And the methods open(), available(), read(), position(), seek().... are used for this purpose. PROGMEM cannot do that. But you can use FFat, LittleFS, SPIFFS with a previously created partition.
